### ISSUE TITLE
thrift: Client and Server interfaces

### ISSUE THREAD

**Issue by abhinav** [yarpc/yarpc-java#247]
Currently the Thrift code we generate exposes the `*_args` and `*_result` objects. This isn't how we want the user experience to be.

Given,

```
service KeyValue {
  void setValue(1: Key key, 2: Value value)
  Value getValue(1: Key key)
    throws (1: ValueDoesNotExist doesNotExist)
}
```

As a service author, I should only implement,

```
class KeyValueImpl implements KeyValue.Handler {
  public CompletionStage<Reply<Void>>
    setValue(Context ctx, Key key, Value value) {
      // ...
  }
  
  public CompletionStage<Reply<Value>> getValue(Context ctx, Key key) {
    return CompletionStage.failedFuture(new ValueDoesNotExist(...));
  }
}
```

How this is registered against the Dispatcher is up for discussion but some ways are,

```
KeyValue.register(dispatcher, new KeyValueImpl());
dispathcer.register(new KeyValueImpl());
dispatcher.register(KeyValue.procedures(new KeyValueImpl());
```

In the background, we'll want to adapt the `KeyValueHandler` into a list of `Procedure`s which hold `UnaryRequestHandler<getValue_args, getValue_result>` and similar.

On the client side, we already expose arguments rather than the `*_args` structs but we still return the `*_result` object which users have to manually interpret the exceptions from. We should change that to always return just the return value (or `Void`) and fail the future with the corresponding exception if an exception was received.

---

To implement this, our yarpc-encoding-thrift code will basically be generic and parameterized over the `_args` and `_result` types but the generated code will adapt that to the user-facing interfaces. This is similar to the approach employed by YARPC Go.

**Comment by bpon** [comment]
We will have to discuss the client side further.  There are a few ways we could handle the response.

1. The current way is to generate a unique `*Reply` type for each procedure that throws thrift exceptions when unwrapping the response data.

```java
client.getValue("foo")
  .call(context)
  .thenAccept(/*GetValueReply*/ reply -> {
    Headers headers = reply.getHeaders();
    // Do something with headers
    try {
      String value = reply.get();
      // Do something with value
    } catch (ValueNotFound exception) {
      // Handle exception
    }
  });
```

Pros:
- Forces user to handle exceptions
- Allows access to headers for both successful and exceptional responses

Cons:
- Requires a unique (but generated) reply type for each procedure
- Users need to call `reply.get()` even for void responses

2. Another option is to handle thrift exceptions with a CompletionStage.

```java
client.getValue("foo")
  .call(context)
  .thenAccept(/* Reply<String> */ reply -> {
    Headers headers = reply.getHeaders();
    // Do something with headers
    String value = reply.get();
    // Do something with value
  })
  .exceptionally(exception -> {
    if (exception instanceof ValueNotFound) {
      ValueNotFound vnf = ValueNotFound.class.cast(exception);
      Headers headers = vnf.getHeaders();
      // Do something with headers
      // Handle vnf
    } else {
      // Handle other exception
    }
    return null;
  });

Pros:
- Utilizes a common `Reply` interface
- Simplifies the CompletionStage for a successful response (`thenAccept`)

Cons:
- Users might forget to handle a specific exception
- Messy exception handling

3. Support both 1 and 2 using different call methods.

```java
client.getValue("foo")
  .call(context)
  .thenAccept(/*GetValueReply*/ reply -> { ... })

client.getValue("foo")
  .callSimple(context)
  .thenAccept(/* Reply<String> */ reply -> { ... })
```

**Comment by abhinav** [comment]
@bpon This was resolved by https://github.com/yarpc/yarpc-java/pull/248 right?

**Comment by bpon** [comment]
Resolved by #248
